rust_src/src/fonts.rs

```diff
@@ -0,0 +1,72 @@
+use remacs_macros::lisp_fn;
+use remacs_sys::{font, Qfont_spec, Qfont_entity, Qfont_object, Qsymbolp, wrong_type_argument};
+use lisp::LispObject;
+use vectors::LispVectorlikeRef;
+
+// A font is not a type in and of itself, it's just a group of three kinds of
+// pseudovector. This newtype allows us to define methods that yield the actual
+// font types: Spec, Entity, and Object.
+pub struct LispFontRef(LispVectorlikeRef);
+
+impl LispFontRef {
+    #[inline]
+    pub fn from_vectorlike(v: LispVectorlikeRef) -> LispFontRef {
+        LispFontRef(v)
+    }
+
+    pub fn is_font_spec(self) -> bool {
+        self.0.pseudovector_size() == font::FONT_SPEC_MAX as i64
+    }
+
+    pub fn is_font_entity(self) -> bool {
+        self.0.pseudovector_size() == font::FONT_ENTITY_MAX as i64
+    }
+
+    pub fn is_font_object(self) -> bool {
+        self.0.pseudovector_size() == font::FONT_OBJECT_MAX as i64
+    }
+}
+
+pub enum FontExtraType {
+    Spec,
+    Entity,
+    Object,
+}
+
+impl FontExtraType {
+    pub fn from_object_or_error(extra_type: LispObject) -> FontExtraType {
+        if extra_type.eq(LispObject::from_raw(unsafe { Qfont_spec })) {
+            FontExtraType::Spec
+        } else if extra_type.eq(LispObject::from_raw(unsafe { Qfont_entity })) {
+            FontExtraType::Entity
+        } else if extra_type.eq(LispObject::from_raw(unsafe { Qfont_object })) {
+            FontExtraType::Object
+        } else {
+            // TODO: This should actually be equivalent to
+            // intern("font-extra-type"), not Qsymbolp.
```

Comment:
In order to call `intern`, you can add it to the imported functions in `remacs-sys`. Signature should be straightforward.